Find the unique positive integer whose square has the form 1_2_3_4_5_6_7_8_9_0,
where each “_” is a single digit.

Only the final few numbers will affect the final few digits. E.g. if we had a number nxxxxx, there is no way that n will affect the final digit of nxxxxx ** 2

So, given our inputs we know that the last digit is "0" because the final digit in the square is "0".

Given that the last digit is "0", we know the last two digits in the square are "00" (because any number ending in zero squared will have two zeroes at the end).

Of the digits remaining in the square, we now need to figure out how to get 9_0. We know that it is actually 900. So the the square % 1000 == 900. In order to get the 900, we need to find a number, n < 10, that when squared % 10 == 9. There are only two available numbers for that: n = 3 (n^2 == 9) or n = 7 (n^2 == 49).

In [6]:
import math
import itertools

In [7]:
original = "1_2_3_4_5_6_7_8_9_0"

there are 9x _ spaces, and so we need 9 variables

In [8]:
num = original

str_range = [str(x) for x in range(10)]
str_range

['0', '1', '2', '3', '4', '5', '6', '7', '8', '9']

In [9]:
max_num = num.replace("_", "9")
mx = int(math.sqrt(int(max_num)))

In [10]:
min_num = num.replace("_", "0")
mn = int(math.sqrt(int(min_num)))

In [11]:
mx - mn

378925613

We know that the smallest number to check would be the squareroot of the number when every _ is replaced with "0".

We know that the largest number to check would be the squareroot of the number when every _ is replaced with "9".

That ends up with 378,925,613 possible digits to check.

In [18]:
(mx - mn) / 50

7578512.26

If we know we only need to check numbers that end in either 30 or 70, that is 2/100 (1/50) numbers we have to check, and so that number reduces our search size to approx 7,578,512 numbers.

In [12]:
for i in range(10, 1010, 10):
    
    
    
    if i % 100 != 30 and i % 100 != 70:
        continue
        
    sq = i**2
    temp = sq // 10000
    temp %= 10
    
    if temp == 8:
        print(f"{i=}")
        print(i**2)
    
        print()

i=430
184900

i=530
280900

i=830
688900



In [14]:
for lead in range(1000, mx, 1000):
    for i in [430]:
        num = lead + i
        temp = num**2 // 10**7
        temp %= 10
        temp2 = num**2 // 10**5
        temp2 %= 10
        if temp == 7 and temp2 == 8:
            print(num)
            print(num ** 2)
            print()

38430
1476864900

131430
17273844900

217430
47275804900

295430
87278884900

388430
150877864900

481430
231774844900

567430
321976804900

645430
416579884900

738430
545278864900

831430
691275844900

917430
841677804900

924430
854570824900

1088430
1184679864900

1181430
1395776844900

1267430
1606378804900

1274430
1624171824900

1531430
2345277844900

1617430
2616079804900

1624430
2638772824900

1881430
3539778844900

1974430
3898373824900

2231430
4979279844900

2324430
5402974824900

2495430
6227170884900

2674430
7152575824900

2845430
8096471884900

2938430
8634370864900

3024430
9147176824900

3195430
10210772884900

3288430
10813771864900

3374430
11386777824900

3467430
12023070804900

3545430
12570073884900

3638430
13238172864900

3724430
13871378824900

3817430
14572771804900

3895430
15174374884900

3988430
15907573864900

4074430
16600979824900

4081430
16658070844900

4167430
17367472804900

4245430
18023675884900

4338430
18821974864900

4431430
19637571844900

45

502845430
252853526471884900

502938430
252947064370864900

503024430
253033577176824900

503195430
253205640772884900

503288430
253299243771864900

503374430
253385816777824900

503467430
253479453070804900

503545430
253558000073884900

503638430
253651668172864900

503724430
253738301378824900

503817430
253832002771804900

503895430
253910604374884900

503988430
254004337573864900

504074430
254091030979824900

504081430
254098088070844900

504167430
254184797472804900

504245430
254263453675884900

504338430
254357251974864900

504431430
254451067571844900

504517430
254537837173804900

504595430
254616547976884900

504688430
254710411375864900

504781430
254804292072844900

504867430
254891121874804900

504945430
254969887277884900

505038430
255063815776864900

505131430
255157761573844900

505217430
255244651575804900

505295430
255323471578884900

505388430
255417465177864900

505481430
255511476074844900

505567430
255598426276804900

505645430
255677300879884900

505738430


896181430
803141155476844900

896267430
803295306078804900

896274430
803307853871824900

896531430
803768604977844900

896617430
803922815779804900

896624430
803935368472824900

896881430
804396299478844900

896974430
804563128073824900

897231430
805024238979844900

897324430
805191132674824900

897495430
805498046870884900

897674430
805819382275824900

897845430
806126416171884900

897938430
806293424070864900

898024430
806447876876824900

898195430
806755030472884900

898288430
806922103471864900

898374430
807076616477824900

898467430
807243722770804900

898545430
807383889773884900

898638430
807551027872864900

898724430
807705601078824900

898817430
807872772471804900

898895430
808012994074884900

898988430
808180197273864900

899074430
808334830679824900

899081430
808347417770844900

899167430
808502067172804900

899245430
808642343375884900

899338430
808809611674864900

899431430
808976897271844900

899517430
809131606873804900

899595430
809271937676884900

899688430


1307324430
1709097165274824900

1307495430
1709544299470884900

1307674430
1710012414875824900

1307845430
1710459668771884900

1307938430
1710702936670864900

1308024430
1710927909476824900

1308195430
1711375283072884900

1308288430
1711618616071864900

1308374430
1711843649077824900

1308467430
1712087015370804900

1308545430
1712291142373884900

1308638430
1712534540472864900

1308724430
1712759633678824900

1308817430
1713003065071804900

1308895430
1713207246674884900

1308988430
1713450709873864900

1309074430
1713675863279824900

1309081430
1713694190370844900

1309167430
1713919359772804900

1309245430
1714123595975884900

1309338430
1714367124274864900

1309431430
1714610669871844900

1309517430
1714835899473804900

1309595430
1715040190276884900

1309688430
1715283783675864900

1309781430
1715527394372844900

1309867430
1715752684174804900

1309945430
1715957029577884900

1310038430
1716200688076864900

1310131430
1716444363873844900

1310217430
1716669713875804900

13102954

In [58]:
digit_list = [ele for ind, ele in enumerate(original) if ind % 2 == 0]

def check_num(n):
    
    checker = 9
    while checker > 1:
        
        n = n // 100
        
        if n % 10 != checker:
            return False
        
        checker -= 1
    
    return True
    

In [60]:
start = mn - (mn % 100)
end = mx

for i in range(start, end, 100):
    for j in [30, 70]:
        num = i + j
        if check_num(num ** 2):
            print(num)



1389019170
